<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера V2</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Ты достаточно невнимательно посмотрела один из красных комментариев первой итерации под функцией прибыли. "Двойной бутстрап" ты устранила, а вот проблема с индексами осталась.

<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера V3</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Проблема осталась, написал комментарий на тему того из-за чего возникает и как устранять.

<div class="alert alert-info">
<font size="4"><b> Комментарий студента_3</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
В твоем коде (аналог функции revenue) я заменила target_subsample (выборка только 500 точек) на первоначальную выборку target_first. После этого, проверила: индексы, которые попали в топ-200 предсказаний попадают в selected_target столько же раз, не задваиваются. Можно и не сбрасывать индексы, чтобы лучше увидеть.

В моей функции когда я исправила код первый раз (убрала бутстрап)  я НЕ передавала отсортированный список индексов лучших предсказаний в сэмпл 500, у меня сразу передавался в исходный таргет: selected = target[predictions_sorted.index][:BEST_POINTS]. Поэтому когда исправляла второй раз и искала баг не смогла найти, что не так в моей функции.
        
Но, я пропустила эту ошибку в твоем коде проверки функции прибыли (см. ниже, я закомментировала предыдующую строку кода, написала новую). 

Как работает код в моих функциях, на мой взгляд и исходя из твоих пояснений, соответствует логики, которую ты описал в своем последнем комментарии: сначала в функции бутстрап мы формируем выборку target на 500 точек. Выборку predictions  формируем исходя из предсказаний на сформированной выборке target на 500 точек. Эти сформированные выборки передаем в функцию расчета прибыли:
1. сортируем  список предсказаний, полученный в функции бутстрап
2. передаем этот отсортированный список предсказаний ограниченный 200 лучшими точками в исходный таргет, в котором индексы не повторяются.   

Надеюсь, мы говорим об одном и том же. 
      </font>
</div>


<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера V4</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Да, описание того, как всё должно работать, абсолютно корректное, с таким алгоритмом ошибки не будет. Более того, есть несколько вариантов как решить проблему, и такой алгоритм я считаю самым удобным и простым.<br>
Сейчас не проверить точно, но допускаю, что я, формируя код на основе написанного тобой, мог случайно одну строчку не скопировать и оставить старую версию. Если так, то прости пожалуйста. Приведи, пожалуйста, текущий код к описанной тобой выше логике (благо для этого надо всего лишь поменять одну переменную на другую) и, кажется, можно принимать работу.<br>
За невнимательность ещё раз прости 😿

<div class="alert alert-info">
<font size="4"><b> Комментарий студента_4</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
    Я поменяла одну переменную в  своей функции бутстрап и  ниже  оставила комментарий с логикой (перед 27 строкой кода).
 </font>
</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
    Привет, Анастасия! Спасибо, что прислала задание :) <br>Меня зовут Никита Сон и я буду проверять твой проект. Предлагаю обращаться друг к другу на ты, как это принято в Практикуме, если ты не против. Но если хочешь на Вы - не буду возражать 🙂
        
Поехали 🚀
    <br />

Мои комментарии обозначены пометкой <b>Комментарий ревьюера</b>. При внесении правок в проект, пожалуйста, не меняй и не удаляй их т.к. дальнейшая проверка будет происходить в том числе на основе того, исправлены замечания в комментариях или нет. Зато ты всегда можешь оставить свои комментарии для меня :)
    </font>
</div>

<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    
В зелёных блоках будут позитивные комментарии
<br />
    </font>
</div>

<div class="alert alert-warning">
<font size="4"><b>⚠️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
В жёлтых - некритичные замечания
    </font>
</div>

<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
В красных - важные замечания, которые надо обязательно устранить
    </font>
</div>

<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
В синих с пиццей - различные нейтральные сообщения, советы на будещее и прочее.
    </font>
</div>

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

## Загрузка и подготовка данных

In [1]:

#импорт библиотек:

import pandas as pd
import seaborn as sns

from numpy.random import RandomState
RANDOM_STATE = RandomState(12345)

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error


In [2]:
#загрузим данные:
data = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
Ну в принципе весьма очевидно, что выше ты делаешь импорты и делаешь загрузку данных. Не стоит оставлять комментари в очевидных местах. Комментарии должны дополнять код, а не дублировать.
    </font>
</div>

In [3]:
# напишем функцию для вывода основной информации о выборке:
def set_inform(df):
    display(df.head(10), df.shape)
    display(df.info())
    display('Количество пропусков:', df.isna().sum())
    display('Количество явных дубликатов:', df.duplicated().sum())
    display('Количество уникальных названий столбца id', df['id'].value_counts().head(15))
    display('Проверка на мультиколлинеарность:', df.corr(method='spearman'))

In [4]:
display('Выборка по первому региону:')
set_inform(data)

'Выборка по первому региону:'

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

(100000, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

'Количество пропусков:'

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

'Количество явных дубликатов:'

0

'Количество уникальных названий столбца id'

TtcGQ    2
A5aEY    2
AGS9W    2
fiKDv    2
HZww2    2
bsk9y    2
74z30    2
Tdehs    2
bxg6G    2
QcMuo    2
At3LI    1
GTrrh    1
0zxMY    1
WPmGa    1
gumu4    1
Name: id, dtype: int64

'Проверка на мультиколлинеарность:'

f0        f1        f2   product
f0       1.000000 -0.471395 -0.002685  0.128417
f1      -0.471395  1.000000  0.001413 -0.181143
f2      -0.002685  0.001413  1.000000  0.486394
product  0.128417 -0.181143  0.486394  1.000000

<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    
Молодец, что проверяешь на мультиколлинеарность 👍<br>

Единственное, в будущем удобнее строить heatmap: по ним легче заметить где наиболее выделяющиеся значения. Особенно когда признаков много.<br>

Что проверила дубликаты - тоже молодец! И что обнаружила, что они как бы не совсем дубликаты - тоже супер.<br>
Что их нельзя удалять - не совсем так, их вполне можно удалить, а можно и нет. В реальном мире надо уточнять у поставщика данных из-за чего такое возможно, и отталкиваться от этого. В нашей ситуацияя, когда мы не можем такое понять, в принципе оба варианта как удалить, так и оставить - корректны.
<br />
    </font>
</div>

In [5]:
display('Выборка по второму региону:')
set_inform(data_1)

'Выборка по второму региону:'

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
5  HHckp  -3.327590  -2.205276  3.003647   84.038886
6  h5Ujo -11.142655 -10.133399  4.002382  110.992147
7  muH9x   4.234715  -0.001354  2.004588   53.906522
8  YiRkx  13.355129  -0.332068  4.998647  134.766305
9  jG6Gi   1.069227 -11.025667  4.997844  137.945408

(100000, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

'Количество пропусков:'

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

'Количество явных дубликатов:'

0

'Количество уникальных названий столбца id'

bfPNe    2
wt4Uk    2
LHZR0    2
5ltQ6    2
Etgwx    1
75SNP    1
1BHQK    1
UcggM    1
Blfu5    1
iXqZP    1
RJosn    1
M2ySO    1
oW7BC    1
xUS2n    1
uFdOH    1
Name: id, dtype: int64

'Проверка на мультиколлинеарность:'

f0        f1        f2   product
f0       1.000000  0.182248 -0.002000 -0.122869
f1       0.182248  1.000000 -0.003678 -0.033908
f2      -0.002000 -0.003678  1.000000  0.975605
product -0.122869 -0.033908  0.975605  1.000000

In [6]:
display('Выборка по третьему региону:')
set_inform(data_2)

'Выборка по третьему региону:'

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
5  LzZXx -0.758092  0.710691  2.585887   90.222465
6  WBHRv -0.574891  0.317727  1.773745   45.641478
7  XO8fn -1.906649 -2.458350 -0.177097   72.480640
8  ybmQ5  1.776292 -0.279356  3.004156  106.616832
9  OilcN -1.214452 -0.439314  5.922514   52.954532

(100000, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

'Количество пропусков:'

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

'Количество явных дубликатов:'

0

'Количество уникальных названий столбца id'

KUPhW    2
VF7Jo    2
xCHr8    2
Vcm5J    2
0BUt9    1
NIWae    1
Zg8eV    1
M4ipZ    1
2tsWS    1
XS37d    1
PFTUn    1
kUcN0    1
O0V3d    1
DZOwT    1
6JDtj    1
Name: id, dtype: int64

'Проверка на мультиколлинеарность:'

f0        f1        f2   product
f0       1.000000  0.002493  0.000053 -0.002464
f1       0.002493  1.000000  0.000378 -0.001463
f2       0.000053  0.000378  1.000000  0.448463
product -0.002464 -0.001463  0.448463  1.000000

In [7]:
#удалим из датасета столбец-индентификатор пробы нефти в каждом регионе, так как он  не представляет ценности для анализа:

df_1 = data.drop(['id'], axis=1)
df_2 = data_1.drop(['id'], axis=1)
df_3 = data_2.drop(['id'], axis=1)

По результатам  подготовки и проверки качества исходных данных выполнено следующее:
- загружены данные геологоразведки трёх регионов. Датасеты содержат 100 000 объектов, пропущенных значений и явных дубликатов нет. Однако, при проверке уникальных значений в столбце id в каждом датасете, найдены одинаковые названия:
 - В первом регионе 10 объектов имеют одинаковые id.
 - Во втором регионе 4 объекта имеют одинаковые id.
 - В третьем регионе 4 объекта имееют одинаковые id.
Ввиду того, что в строках с одинаковыми id параметры в остальных столбцах различаются, это не дубликаты, поэтому удалять их нельзя. Повторы в названиях могут быть связаны с ошибками рандом названия. С учетом их незначительного количества в каждом датасете принято решение оставить их в выборках без изменения.
У каждого объекта 4 признака, из них признак  объём запасов в скважине (product) является целевым, который нужно будет предсказать.
- Удален из выборок столбец id, содержащий данные типа object, так как содержащиеся в нем данные не представляют ценности для дальнейшего анализа.
- Проверка на мультиколлинеарность показала наличие высокой кореляции (0,97) во втором датасете между целевым признаком product и параметром f2. В первом и третьем датасетах высокой корреляции между признаками нет.
- Целевой признак является количественным поэтому необходимо будет решать задачу регрессии.


<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />  
Единственное, чего тут не хватает: гистограмм признаков. По ним можно заметить, что разброс признака f2 в первом регионе заметно больше, чем двух других, почти на порядок (в 10 раз), а так как у нас заранее известно, что модель линейная, обучение которой от масштаба признаков сильно зависит, есть смысл произвести масштабирование признаков.<br>

Но ты ниже всё равно делаешь масштабирование, так что в целом всё очень даже хорошо.
    </font>
</div>

## Обучение и проверка модели

### Подготовка данных для машинного обучения

In [8]:
# запишем целевой признак и остальные признаки в переменные:
# для первого региона:
target_first = df_1['product']
features_first= df_1.drop(['product'], axis=1)

# для второго региона:
target_second = df_2['product']
features_second= df_2.drop(['product'], axis=1)

# для третьего региона:
target_third = df_3['product']
features_third= df_3.drop(['product'], axis=1)


In [9]:
# разделим  каждую выборку на обучающую и валидационную в соотношении 75:25:
# для первого региона:

features_first_train, features_first_valid, target_first_train, target_first_valid = train_test_split(features_first, target_first, test_size=0.25, random_state=12345) 
print('Первый регион:\n','тренировочная выборка:', features_first_train.shape, 'валидационная выборка', features_first_valid.shape )
# для второго региона:
features_second_train, features_second_valid, target_second_train, target_second_valid = train_test_split(features_second, target_second, test_size=0.25, random_state=12345) 
print('Второй регион:\n','тренировочная выборка:', features_second_train.shape, 'валидационная выборка', features_second_valid.shape )
# для третьего региона:
features_third_train, features_third_valid, target_third_train, target_third_valid = train_test_split(features_third, target_third, test_size=0.25, random_state=12345) 
print('Третий регион:\n','тренировочная выборка:', features_third_train.shape, 'валидационная выборка', features_third_valid.shape )

Первый регион:
 тренировочная выборка: (75000, 3) валидационная выборка (25000, 3)
Второй регион:
 тренировочная выборка: (75000, 3) валидационная выборка (25000, 3)
Третий регион:
 тренировочная выборка: (75000, 3) валидационная выборка (25000, 3)


<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />

Аналогично с логикой почему повторяющийся код стоит оформлять в функции, константные значения, которые повторяются в коде в нескольких местах, такие как random state или test size, стоит заносить в переменные. Представь ты сделал работу по анализу 50 регионов, а потом заказчик попросил поменять размер теста с 0.25 на 0.3, решив, что оценка на 0.25 недостаточно точная. Менять число в 50 местах кода - куда хуже, чем в начале кода поменять TEST_SIZE = 0.25 на TEST_SIZE = 0.3. Код становится намного легче поддерживать, говоря профессиональным языком.

### Обучение модели

In [10]:
# проведем нормализацию признаков:
def scaler_sets(features_train, features_valid):

    numeric = ['f0','f1','f2']

    scaler = StandardScaler()
    scaler.fit(features_train[numeric])
    pd.options.mode.chained_assignment=None
    features_train[numeric] = scaler.transform(features_train[numeric])
    features_valid[numeric] = scaler.transform(features_valid[numeric])

    return features_train, features_valid
    


In [11]:
# масштабируем признаки первой выборки:
scaler_sets(features_first_train, features_first_valid)

(             f0        f1        f2
 27212 -0.544828  1.390264 -0.094959
 7866   1.455912 -0.480422  1.209567
 62041  0.260460  0.825069 -0.204865
 70185 -1.837105  0.010321 -0.147634
 82230 -1.299243  0.987558  1.273181
 ...         ...       ...       ...
 4094   1.567114 -1.087243 -0.272211
 85412 -1.904207 -0.525360  1.327530
 2177   0.418949 -1.296788 -0.196407
 77285  0.400077 -1.466874 -0.445317
 86498  1.746246  0.027415  2.766848
 
 [75000 rows x 3 columns],
              f0        f1        f2
 71751  0.517917 -0.610097 -0.126226
 80493  0.568391 -0.086063 -0.814914
 2655   0.805688 -1.613289 -1.341342
 53233  0.222503 -1.355437 -0.597275
 91141 -0.087941  1.433113 -0.544588
 ...         ...       ...       ...
 12581  1.541222 -0.709525  0.024535
 18456 -0.522565  0.474447 -0.103985
 73035 -0.138850  1.595071 -0.903485
 63834  0.658267 -1.428224  0.621967
 43558  1.409556 -0.650407  0.769670
 
 [25000 rows x 3 columns])

In [12]:
# масштабируем признаки второй выборки:
scaler_sets(features_second_train, features_second_valid)

(             f0        f1        f2
 27212 -0.850855  0.624428  0.296943
 7866   1.971935  1.832275  0.294333
 62041  1.079305  0.170127 -0.296418
 70185 -1.512028 -0.887837 -0.880471
 82230 -1.804775 -0.718311 -0.293255
 ...         ...       ...       ...
 4094  -0.334573 -0.005156  0.295560
 85412  0.415280 -0.249559  0.877957
 2177   1.047070 -0.649992  1.473368
 77285 -0.114780 -1.190699  0.299156
 86498 -0.646146  0.099075  0.295611
 
 [75000 rows x 3 columns],
              f0        f1        f2
 71751 -0.168616  0.571372  0.297471
 80493  0.878384 -1.772903 -0.293901
 2655  -0.852976 -0.004986 -0.876502
 53233  1.496940 -1.146077 -0.294011
 91141  0.552177  0.057961 -1.457764
 ...         ...       ...       ...
 12581 -0.585274  0.036061  1.465670
 18456 -1.059992 -0.305376  0.880934
 73035 -1.146172 -0.119247  1.471417
 63834 -0.839466  0.785521  0.302943
 43558  1.048780 -0.275907 -0.295706
 
 [25000 rows x 3 columns])

In [13]:
# масштабируем признаки третьей выборки:
scaler_sets(features_third_train, features_third_valid)

(             f0        f1        f2
 27212 -0.526160  0.776329 -0.400793
 7866  -0.889625 -0.404070 -1.222936
 62041 -1.133984  0.208576  0.296765
 70185  1.227045  1.570166 -0.764556
 82230 -0.194289  0.878312  0.840821
 ...         ...       ...       ...
 4094  -1.815240  0.398261  1.821017
 85412 -0.274576 -1.265211  0.245519
 2177   0.368564  0.797225  0.665575
 77285 -2.440690  0.113784  0.440907
 86498 -1.732468  0.393573 -1.424465
 
 [75000 rows x 3 columns],
              f0        f1        f2
 71751 -0.836717 -2.227439 -0.078489
 80493 -0.821648  0.740237 -1.000811
 2655  -2.651336 -0.235721 -0.244546
 53233  1.077998  0.938090  0.511313
 91141 -1.173938  2.387270  1.034363
 ...         ...       ...       ...
 12581 -0.761845 -0.067861 -0.819227
 18456  0.006155  0.509468  0.030344
 73035  0.416444 -0.084595  0.218987
 63834 -0.136658 -0.101235 -0.870787
 43558 -1.359565 -0.598457  1.716644
 
 [25000 rows x 3 columns])

<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />

Интересный подход в масштабированию! Сходу даже не могу сказать насколько хороший, но точно необычный! А необычные шаги студентов - это всегда круто! :)

In [14]:
# напишем функцию обучения модели с помощью  линейной регрессии:
def linear_regression(features_train, target_train, features_valid, target_valid):
    model= LinearRegression()
    model.fit(features_train, target_train) 
    predictions = pd.Series(model.predict(features_valid), index=features_valid.index)
          
    print("Корень из средней квадратичной ошибки(RMSE) =", (mean_squared_error(target_valid, predictions) ** 0.5))
            
#  найдем RMSE для константной модели: каждому объекту она прогнозирует среднее значение целевого признака (средний запас)
     
    predicted_valid_mean =pd.Series(target_train.mean(), index=target_valid.index) 
    print("RMSE константной модели", (mean_squared_error(target_valid, predicted_valid_mean) ** 0.5))
    
    #посчитаем средний запас предсказанного сырья:
    print ("Средний запас предсказанного сырья", predictions.mean())
    
    
    return target_valid, predictions
    
 


In [15]:
# при помощи функции выведем на экран основные метрики модели линейной регрессии
#сохраним предсказания и правильные ответы на валидационной выборке для каждого региона:
display('Для первого региона:')
first_target, first_predictions = linear_regression(features_first_train, target_first_train, features_first_valid, target_first_valid)


'Для первого региона:'

Корень из средней квадратичной ошибки(RMSE) = 37.5794217150813
RMSE константной модели 44.289591053907365
Средний запас предсказанного сырья 92.59256778438035


<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
В реальной задаче у тебя может быть не 3, а 50 регионов, так что лучше называть не first_target, а target_1 и аналогично.

In [16]:
display('Для второго региона:')
second_target, second_predictions = linear_regression(features_second_train, target_second_train, features_second_valid, target_second_valid)

'Для второго региона:'

Корень из средней квадратичной ошибки(RMSE) = 0.893099286775617
RMSE константной модели 46.02144533725462
Средний запас предсказанного сырья 68.728546895446


In [17]:
display('Для третьего региона:')
third_target, third_predictions = linear_regression(features_third_train, target_third_train, features_third_valid, target_third_valid)

'Для третьего региона:'

Корень из средней квадратичной ошибки(RMSE) = 40.02970873393434
RMSE константной модели 44.90234968510566
Средний запас предсказанного сырья 94.96504596800489


<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    

Всё супер: требования заказчика по выбору модели и по балансу обучающей тестовой выборки соблюдены, модели построены корректно, их предсказания адекватны.


Для каждого региона проведена нормализация признаков и построена модель линейной регрессии.
Проведена проверка качества полученных предсказаний на валидационных выборках при помощи метрики RMSE. Чтобы оценить адекватность полученной модели проведено сравнение c RMSE константной модели,  в которой  каждому объекту она прогнозирует среднее значение целевого признака (средний запас сырья).
По результатам сравнения полученных оценок установлено следующее:
- для первого региона RSME на валидационной выборке меньше RMSE константной модели на 6,7 тыс.барр.
- для второго региона RSME на валидационной выборке меньше RMSE константной модели на 45,1 тыс.барр.
- для третьего региона RSME на валидационной выборке меньше RMSE константной модели на 4,8 тыс.барр.

Таким образом, наилучшее качество модели получено на валидационных данных второго региона так как получено минимальное значение метрики RMSE.

Средний запас предсказанного сырья в первом регионе равен 92,6 тыс. баррелей, во втором регионе - 68,7 тыс. баррелей, в третьем регионе - 95 тыс. баррелей. Среднее количество сырья примерно находится на одном уровне в 1 и 3 регионе и больше в 1.4 раза, чем во 2 регионе.

## Подготовка к расчёту прибыли

In [18]:
# для расчета прибыли создадим константы:
# количество исследуемых точек при разведке
POINTS = 500 

# количество лучших исследуемых точек 
BEST_POINTS = 200 

# бюджет на разработку скважин в регионе
BUDGET = 1e10

# цена за 1000 баррелей сырья
PRODUCT_PRICE = 450000

# максимальная вероятность убытков
MAX_LOST = 0.025

# количество выборок для bootstrap 
BOOTSTRAP_SAMPLES = 1000 



<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
Ну, 10 * 1e9 можно записать как 1e10, но ладно, не страшно )

In [19]:
#рассчитаем минимальное среднее количество продукта в месторождениях региона, достаточное для разработки:
# рассчитаем бюджет на разработку одной скважины:
budget_point = BUDGET / BEST_POINTS

# рассчитаем минимальный объем сырья для одной скважины:
min_product = budget_point / PRODUCT_PRICE
print('Минимальный объём сырья для безубыточной разработки новой скважины (тыс. баррелей): {:.2f}'.format(min_product))

Минимальный объём сырья для безубыточной разработки новой скважины (тыс. баррелей): 111.11


Рассчитанный минимальный объем сырья для безубыточной разработки одной новой скважины равен 111.11 тыс. баррелей и превышает средний запас предсказанного сырья в трех регионах. 

<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    

Расчёт корректен. Молодец, что обрезала дробь до 2 знаков после запятой. Единственное, стоит и в print добавлять единицы измерения. Но тк написала какие единицы в выводе - не страшно.

## Расчёт прибыли по выбранным скважинам и предсказаниям модели

<div class="alert alert-warning">
<font size="4"><b>⚠️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
Названия функций и переменных должны отражать их смысл. Ты решаешь задачу регрессии, у тебя нет вероятностей (probabilities).

<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Во-первых, функция расчёта прибыли должна вычислять прибыль по выбраным скважинам, а не делать бутстрап. Ты же делаешь бустрап ниже, в другой функции, там же вызываешь данную функцию revenue, и она снова делает сэмпл. Это неправильно, так быть не должно.
    </font>
</div>

<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Во-вторых, в этой задаче надо быть очень аккуратной с индексами. Я ниже проделал операции, которые ты делаешь сначала в функции бутстрапа, потом - в ревенью. Видно, что количество элементов с повторяющими индексами различается в топ-предсказаниях и соответствующих таргетах. Кроме того, в том подмножестве таргетов, по которому ты считаешь прибыль, оказывается более 200 элементов. Очевидно, так не должно быть. Когда уберёшь лишнее сэмплирование, проверь, что эти моменты будут корректны и, если нет, исправляй баг.
    </font>
</div>

<div class="alert alert-info">
<font size="4"><b> Комментарий студента</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Функция расчета прибыли исправлена.
    </font>
</div>


<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера V3</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
К сожалению, это не поможет потому что ты ведь сбросила индексы уже после того, как отобрала 200 скважин для вычисления прибыли, так что индексы другие, а результат математически тот же самый.<br>

Смотри, проблема возникает из-за двух факторов:<br>
1. В семпле есть повторяющиеся индексы. Это можно устранить сделав резет индексов, но не после того, как ты отобрала 200 скважин для расчёта прибыли, а до.<br>
2. Ты сначала по индексам одного сериес берёшь кусок другого, а потом снова по индексам берёшь кусок уже вот этого куска. Это починить ещё проще: нужно отобрав 200 лучших по предсказаниям скважин брать их индексы и передавать не в сэмпл таргета из 500 элементов, а в весь таргет (в нём то индексы не повторяются)<br>

Надеюсь сейчас стало чуть понятнее. Если ещё нет, пиши об этом комментарий, разберём ещё поглубже.

<div class="alert alert-danger">
<font size="4"><b>❌ Комментарий ревьюера V2</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
Ты очень зря просто взяла и закомментировала весь мой код. Я же не просто так его написал, а понимал, что у тебя есть баг, который надо исправить. Стоило привести код ниже в соответствии с тем, как ты поменяла расчёт прибыли и проверить всё ли корректно.<br>

Я привёл всё к текущему виду и запустил. Какой результат мы видим

<div class="alert alert-info">
<font size="4"><b> Комментарий студента_2</b></font>
    <br /> 
    <font size="3", color = "black">
<br /> 
не знаю, правильно или нет, применила reset_index к selected_target. Других идей нет как это поправить.
    </font>
</div>


In [20]:
# КОД РЕВЬЮЕРА V2
# аналог функции bootstrap
target_subsample = first_target.sample(POINTS, replace=True, random_state=RANDOM_STATE)
predictions_subsample = first_predictions[target_subsample.index]

display(target_subsample)
print('-'*35)
display(predictions_subsample)
print('='*35)
display(target_subsample.index.value_counts())
print('-'*35)
display(predictions_subsample.index.value_counts())

13616     34.280424
84656     55.567766
73037     26.938283
19345    133.142251
68339     12.386952
            ...    
13213     90.511196
67359    134.992329
71785     74.874896
94411     90.708239
87014    146.577081
Name: product, Length: 500, dtype: float64

-----------------------------------


13616     69.031631
84656    108.084038
73037     89.065885
19345    109.885938
68339     71.893169
            ...    
13213     51.883205
67359     88.959983
71785    113.393785
94411     54.043185
87014    115.134762
Length: 500, dtype: float64

13642    2
55668    2
38891    2
30480    2
46965    2
        ..
73037    1
79179    1
78782    1
74055    1
77823    1
Length: 493, dtype: int64

-----------------------------------


13642    2
55668    2
38891    2
30480    2
46965    2
        ..
73037    1
79179    1
78782    1
74055    1
77823    1
Length: 493, dtype: int64

<div class="alert alert-danger">
    <font size="3", color = "black">
После бутстрапа всё вроде сходится, индексы одинаковые в предсказаниях и таргетах. Некоторые встречаются больше 1 раза, но это нормально, так и работает бутстрап, что набирает подвыборку случайным образом из элементов исходных данных, с возможными повторами. Так что встретившиеся у нас повторяющиеся индексы - это индексы тех элементов, которые благодаря бутстрапу попали в нашу подвыборку больше 1 раза.<br>
А теперь сделаем над этими Series операции, которые ты делаешь в вычислении прибыли, и посмотрим что получается.

In [21]:
# КОД РЕВЬЮЕРА V2
# аналог функции revenue
#research_points =probs_subsample.sample(POINTS, replace=True, random_state=RANDOM_STATE)

predictions_sorted = predictions_subsample.sort_values(ascending = False)

#selected_target = target_subsample[predictions_sorted.index][:BEST_POINTS]

#исправила target_subsample на first_target:

selected_target = first_target[predictions_sorted.index][:BEST_POINTS]

display(predictions_sorted[:BEST_POINTS].index.value_counts())
print('-'*35)

display(selected_target.index.value_counts())
print('-'*35)
print('Количество таргетов, по которым считается прибыль:', selected_target.shape[0])





13642    2
23465    2
30480    2
55668    2
2556     1
        ..
82431    1
56158    1
71775    1
34146    1
61183    1
Length: 196, dtype: int64

-----------------------------------


13642    2
23465    2
30480    2
55668    2
2556     1
        ..
82431    1
56158    1
71775    1
34146    1
61183    1
Length: 196, dtype: int64

-----------------------------------
Количество таргетов, по которым считается прибыль: 200


<div class="alert alert-danger">
    <font size="3", color = "black">
Индексы, которые попали в топ-200 предсказаний попали в selected_target не 2 раза, а 4!<br>

Почему так получилось? Смотри, после того как ты сделал бутстрапированную выборку предсказаний, ты по их индексам берёшь соответствующую часть таргетов. Т.к. у нас в бутстрапе могут повторяться элементы, то несколько индексов у нас теперь есть совпадающие (например, 14512).<br>
Затем ты сортируешь значения предсказаний, берёшь первые 200 индексов и по ним берёшь соответствующие значения таргетов. И вот тут то и кроется дьявол! Есть у тебя две записи с индексом 14512 в сериес предсказаний, и, логично, две записи с индексом 14512 в сериес с таргетом. Индекс 14512 попадает в итоговый список лучших предсказаний, ты говоришь пандасу: "пандас, а возьми-ка мне для расчёта прибыли из таргета строку с индексом 14512." Послушный пандас идёт брать эту строку и видит, что у тебя таких в сериес с таргетами вообще-то две штуки. Как ты думаешь, какую из них он возьмёт? Правильно, обе. Легко заметить, что в predictions_sorted.index у тебя 14512 встречается 2 раза, а в том, по которому ты считаешь прибыль, 4. Потому что ты два раза попросил его взять 14512, ну он и взял 2 раза, правда оба раза он взял обе строки таргета с индексом 14512, и их получилось 4 :)<br>

Я понимаю, что дедлайны горят, но очень надеюсь, что ты поняла из-за чего возникает баг и как его исправить. Крайне желательно, чтобы ты устранила сама. Надо же развивать твой скилл, специалиста готовим :)<br>
Не сможешь или всё ещё непонятно почему индексы разошлись - оставь комментарий об этом, я подскажу.<br>

После исправления запусти весь код, чтобы убедиться, что всё работает корректно, при необходимости, поправь выводы.

In [22]:
#Напишем функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели:
def revenue (predictions, target):
    
    predictions_sorted = predictions.sort_values(ascending = False)
    selected = target[predictions_sorted.index][:BEST_POINTS]
    revenue = (selected.sum() * PRODUCT_PRICE) - BUDGET
    
    return (revenue / 1e6).round(2)

In [23]:
# применим функцию на данных в каждом регионе:
display('Для первого региона:')
print ('Прибыль (млн. руб):', revenue(first_predictions, first_target))

'Для первого региона:'

Прибыль (млн. руб): 3320.83


In [24]:
display('Для второго региона:')
print ('Прибыль (млн. руб):', revenue(second_predictions, second_target))

'Для второго региона:'

Прибыль (млн. руб): 2415.09


In [25]:
display('Для третьего региона:')
print ('Прибыль (млн. руб):', revenue(third_predictions, third_target))

'Для третьего региона:'

Прибыль (млн. руб): 2710.35


Для расчета прибыли были выбраны 200 точек с максимальными значениями предсказанного объема запасов сырья. Просуммировав  целевое значение объема сырья, соответствующее предсказанным, получено значение прибыли.


## Расчёт рисков и средней прибыли для каждого региона

In [26]:
def revenue_bootstrap(predictions, target):
    
    values = []
    for i in range(BOOTSTRAP_SAMPLES):
        target_subsample = target.sample(POINTS, replace=True, random_state=RANDOM_STATE)
        predictions_subsample = predictions[target_subsample.index]
        revenue_sample = revenue(predictions_subsample, target)
        values.append(revenue_sample)
    
    values = pd.Series(values)
    mean = values.mean()
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    print('Средняя прибыль (млн. руб): {:.2f}'.format(mean))
    print('95% доверительный интервал (млн. руб): ({:.2f}, {:.2f})'.format(lower, upper))
       

    
    lost_predict = (values <0).sum() / BOOTSTRAP_SAMPLES
    
    if lost_predict < MAX_LOST: 
        print('Вероятность убытков равна {}, что меньше чем максимальная допустимая вероятность {}'.format(lost_predict, MAX_LOST))
    else:
        print('Вероятность убытков равна {}, что больше чем максимальная допустимая вероятность {}'.format(lost_predict, MAX_LOST))    

    return mean, lower, upper, lost_predict
    
    

<div class="alert alert-info">
<font size="4">🍕<b> Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
В функции бутстрапа всё кажется корректным. Но с багом в расчёте прибыли, конечно, не можем считать итоговые выводы состоятельными. Так что жду вторую отправку с исправлениями!

<div class="alert alert-info">
<font size="4"><b> Комментарий студента_4</b></font>
    <br /> 
    <font size="3", color = "black">
<br />
Логика работы  функции бутстрапа:
        
- формируем выборку target на 500 точек и на ней получаем предсказания (это 2 первые строки в цикле)
- далее вызываем функцию расчета прибыли.

Но в функции расчета прибыли мы должны взять предсказания полученные на выборке target 500 и отсортировать и передать их во всю выборку target, выбирая 200 лучших. 
        
Поэтому в третьей строке цикла функции бутстрап, где я вызываю функцию расчета прибыли в качестве параметров я указываю  predictions_subsample и target, а target_subsample как у меня было ранее.

 

In [27]:
overall = []

In [28]:
overall.append(revenue_bootstrap(first_predictions, first_target))

Средняя прибыль (млн. руб): 395.53
95% доверительный интервал (млн. руб): (-111.21, 909.76)
Вероятность убытков равна 0.07, что больше чем максимальная допустимая вероятность 0.025


In [29]:
overall.append(revenue_bootstrap(second_predictions, second_target))

Средняя прибыль (млн. руб): 461.17
95% доверительный интервал (млн. руб): (78.05, 862.95)
Вероятность убытков равна 0.007, что меньше чем максимальная допустимая вероятность 0.025


In [30]:
overall.append(revenue_bootstrap(third_predictions, third_target))

Средняя прибыль (млн. руб): 392.58
95% доверительный интервал (млн. руб): (-112.23, 934.56)
Вероятность убытков равна 0.065, что больше чем максимальная допустимая вероятность 0.025


In [31]:
overall_df = pd.DataFrame(overall, columns = ['revenue_mean', 'lower', 'upper', 'lost_predict'])
overall_df.head()

revenue_mean      lower      upper  lost_predict
0     395.52562 -111.21425  909.76425         0.070
1     461.16782   78.05400  862.95000         0.007
2     392.57861 -112.22625  934.56250         0.065

В соответствии с полученными статистическими характеристиками можно сделать вывод, что вероятность убытков ниже, чем максимальная допустимая вероятность только во втором регионе (вторая строка итоговой таблицы).Во втором регионе самый лучший узкий доверительный интервал и средняя прибыль, поэтому для разработки скважин должен быть выбран второй регион. 


<div class="alert alert-success">
<font size="4"><b>✔️ Комментарий ревьюера</b></font>
    <br /> 
    <font size="3", color = "black">
<br />    
Расчёты прибыли и доверительных интервалов корректны. На основе корректных вычислений выбран регион для разработки мест нефтедобычи и выбор региона корректно обоснован. Кажется, у заказчика нет оснований не довериться этому экспертному исследованию! 🙂

## Общий вывод

1.По результатам подготовки и проверки качества исходных данных выполнено следующее:

загружены данные геологоразведки трёх регионов. Датасеты содержат 100 000 объектов, пропущенных значений и явных дубликатов нет. Однако, при проверке уникальных значений в столбце id в каждом датасете, найдены одинаковые названия:
- в первом регионе 10 объектов имеют одинаковые id.
- во втором регионе 4 объекта имеют одинаковые id.
- в третьем регионе 4 объекта имеют одинаковые id. Ввиду того, что в строках с одинаковыми id параметры в остальных столбцах различаются, это не дубликаты, поэтому удалять их нельзя. Повторы в названиях могут быть связаны с ошибками рандом названия. С учетом их незначительного количества в каждом датасете принято решение оставить их в выборках без изменения. У каждого объекта 4 признака, из них признак объём запасов в скважине (product) является целевым, который нужно будет предсказать.
Удален из выборок столбец id, содержащий данные типа object, так как содержащиеся в нем данные не представляют ценности для дальнейшего анализа.
Проверка на мультиколлинеарность показала наличие высокой корреляции (0,97) во втором датасете между целевым признаком product и параметром f2. В первом и третьем датасетах высокой корреляции между признаками нет.
Целевой признак является количественным поэтому необходимо будет решать задачу регрессии.

2. Для каждого региона проведена нормализация признаков и построена модель линейной регрессии.
Проведена проверка качества полученных предсказаний на валидационных выборках при помощи метрики RMSE. Чтобы оценить адекватность полученной модели проведено сравнение c RMSE константной модели,  в которой  каждому объекту она прогнозирует среднее значение целевого признака (средний запас сырья).
По результатам сравнения полученных оценок установлено следующее:
- для первого региона RSME на валидационной выборке меньше RMSE константной модели на 6,7 тыс.барр.
- для второго региона RSME на валидационной выборке меньше RMSE константной модели на 45,1 тыс.барр.
- для третьего региона RSME на валидационной выборке меньше RMSE константной модели на 4,8 тыс.барр.

Таким образом, наилучшее качество модели получено на валидационных данных второго региона, так как получено минимальное значение метрики RMSE.

Средний запас предсказанного сырья в первом регионе равен 92,6 тыс. баррелей, во втором регионе - 68,7 тыс. баррелей, в третьем регионе - 95 тыс. баррелей. Среднее количество сырья примерно находится на одном уровне в 1 и 3 регионе и больше в 1.4 раза, чем во 2 регионе.

3. Рассчитан минимальный объем сырья для безубыточной разработки одной новой скважины и  равен 111.11 тыс. баррелей, что превышает средний запас предсказанного сырья в трех регионах.

4. Для расчета прибыли были взяты  200 точек с максимальными значениями предсказанного объема запасов сырья. Просуммировав  целевое значение объема сырья, соответствующее предсказанным, получено значение прибыли. Самое высокое значение прибыли получено для скважин с максимальными запасами сырья третьего региона.

5. Проведен расчет рисков и средней прибыли для каждого региона с применением техники Bootstrap. Полученное значение средней прибыли после оценки рисков намного меньше значения прибыли, полученной для запаса сырья, рассчитанного для 200 скважин с максимальными значениями предсказаний.

6. В соответствии с полученными статистическими характеристиками можно сделать вывод, что вероятность убытков ниже, чем максимальная допустимая вероятность только во втором регионе (вторая строка итоговой таблицы). Во втором регионе самый лучший узкий доверительный интервал и средняя прибыль, поэтому для разработки скважин должен быть выбран второй регион. 


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован